# In this task, you will be working on the Pima Indians Diabetes Database.



Your task is find the optimal hyperparameter value using which you can perform the best classification using GridSearchCV

In [1]:
import numpy as np
import pandas as pd

In [2]:
data=pd.read_csv("diabetes.csv")

In [3]:
data.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

#  Splitting the dataset in X and y

In [4]:
X=data.iloc[:,0:7].values
y=data.iloc[:,-1].values

In [5]:
X.shape

(768, 7)

In [6]:
y.shape

(768,)

In [7]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=1)

# Applying Algorithm to train the model

In [8]:
from sklearn.tree import DecisionTreeClassifier
clf=DecisionTreeClassifier()

In [9]:
clf.fit(X_train,y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

# Prediction without defining any HyperParameter 

In [10]:
y_pred= clf.predict(X_test)

In [11]:
from sklearn.metrics import accuracy_score

In [12]:
accuracy_score(y_test,y_pred)

0.6493506493506493

# Prediction with HyperParameter

In [13]:
param_dist={
    "criterion":["gini","entropy"],
    "max_depth":[1,2,3,4,5,6,7,8,None]
}

In [14]:
from sklearn.model_selection import GridSearchCV
grid= GridSearchCV(clf,param_grid=param_dist,cv=10,n_jobs=-1)

In [15]:
grid.fit(X_train,y_train)

GridSearchCV(cv=10, error_score='raise',
       estimator=DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best'),
       fit_params=None, iid=True, n_jobs=-1,
       param_grid={'criterion': ['gini', 'entropy'], 'max_depth': [1, 2, 3, 4, 5, 6, 7, 8, None]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [16]:
grid.best_estimator_

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=3,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [17]:
grid.best_score_

0.7361563517915309

In [18]:
grid.best_params_

{'criterion': 'entropy', 'max_depth': 3}

# Conclusion:
   From this we can say that by defining two Hyper parameters 'criterion':'entropy' and 'max_depth':3 we can get more accurate      predicition for our dataset in this Machine Learning model
   
   As the predicition without Hyper Parameter is: 0.7077922077922078
   
   And the prediction with Hyper Parameter is: 0.7361563517915309 
   
   The best method for criterion is : entropy
   
   The best value for max_depth is : 3 